File: src/lib/operations/userFunction.js
/**
* @copyright 2023 - Max Bebök
* @license Apache-2.0
*/
import state from "../state.js";
import {asm, asmFunction, asmNOP} from "../intsructions/asmWriter.js";
import opsScalar from "./scalar.js";

/**
 * @param {string} name
 * @param {ASTFuncArg[]} args
 * @returns {ASM[]}
 */
export function callUserFunction(name, args)
{
  let userFunc = state.getFunction(name);
  if(!userFunc) {
    if(!state.varExists(name)) {
      state.throwError("Function "+name+" not known!");
    }
    userFunc = {
      name: state.getVarReg(name),
      args: [],
      isRelative: false
    };
  }

  const res = [];

  if(userFunc.args.length !== args.length) {
    state.throwError(`Function ${name} expects ${userFunc.args.length} arguments, got ${args.length}!`, args);
  }

  for(let i = 0; i < args.length; ++i) {
    const argUser = args[i];
    const argDef = userFunc.args[i];
    if(argUser.type === "num") {
      res.push(...opsScalar.loadImmediate(argDef.reg, argUser.value));
    } else {
      const argVar = state.getRequiredVar(argUser.value, "arg" + i);
      if(argVar.type !== argDef.type) {
        state.throwError(`Function ${name} expects argument ${i} to be of type ${argDef.type}, got ${argVar.type}!`, argUser);
      }
      if(argVar.reg !== argDef.reg) {
        state.throwError(`Function ${name} expects argument ${i} to be in register ${argDef.reg}, got ${argVar.reg}!`, argUser);
      }
    }
  }
  let isRelative = state.getAnnotations("Relative").length > 0;
  if(userFunc.isRelative) {
    isRelative = true;
  }

  const regsArg = userFunc.args.map(arg => arg.reg);
  res.push(asmFunction(userFunc.name, regsArg, isRelative), asmNOP());
  return res;
}

File: cpp/src/operations/user_function.cpp
#include "user_function.h"
#include "scalar.h"

#include "../asm.h"
#include "../state.h"

namespace rspl::ops {

std::vector<AsmInst> callUserFunction(
    const std::string &name, const std::vector<ast::FuncArg> &args) {
  const FuncDef *userFunc = state.getFunction(name);
  if (!userFunc) {
    if (!state.varExists(name)) {
      state.throwError("Function " + name + " not known!");
    }
    // Indirect call through register variable
    static FuncDef indirect;
    indirect.name = *state.getVarReg(name);
    indirect.isRelative = false;
    userFunc = &indirect;
  }

  std::vector<AsmInst> res;

  if (userFunc->args.size() != args.size()) {
    state.throwError("Function " + name + " expects " +
                     std::to_string(userFunc->args.size()) +
                     " arguments, got " +
                     std::to_string(args.size()) + "!");
  }

  for (size_t i = 0; i < args.size(); ++i) {
    const auto &argUser = args[i];
    const auto &argDef = userFunc->args[i];
    if (argUser.type == "num") {
      auto load =
          loadImmediate(argDef.reg, argUser.value);
      res.insert(res.end(), load.begin(), load.end());
    } else {
      const VarDef *argVar =
          state.getRequiredVar(argUser.value, "arg" + std::to_string(i));
      if (argVar->type != argDef.type) {
        state.throwError("Function " + name +
                         " expects argument " + std::to_string(i) +
                         " to be of type " + argDef.type +
                         ", got " + argVar->type + "!");
      }
      if (argVar->reg != argDef.reg) {
        state.throwError("Function " + name +
                         " expects argument " + std::to_string(i) +
                         " to be in register " + argDef.reg +
                         ", got " + argVar->reg + "!");
      }
    }
  }

  bool isRelative = userFunc->isRelative;
  auto annos = state.getAnnotations("Relative");
  if (!annos.empty()) isRelative = true;

  std::vector<std::string> regsArg;
  for (const auto &arg : userFunc->args) {
    regsArg.push_back(arg.reg);
  }

  res.push_back(asmFunction(userFunc->name, regsArg, isRelative));
  res.push_back(asmNOP());
  return res;
}

} // namespace rspl::ops
